msal/wam.py

```diff
@@ -0,0 +1,123 @@
+        self.signal.set()
+        self.auth_result = auth_result
+
+
+def _read_account_by_id(account_id):
+    callback_data = _CallbackData()
+    pymsalruntime.read_account_by_id(
+        account_id,
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+def _signin_silent():
+    callback_data = _CallbackData()
+    pymsalruntime.signin_silent(
+        # TODO: Add other input parameters
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+def _signin_interactive():
+    callback_data = _CallbackData()
+    pymsalruntime.signin_interactive(
+        # TODO: Add other input parameters
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+def _acquire_token_silently(authority, client_id, account):
+    params = pymsalruntime.MSALRuntimeAuthParameters(client_id, authority)
+    callback_data = _CallbackData()
+    pymsalruntime.signin_interactive(
+        params,
+        "correlation", # TODO
+        account,
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+def _acquire_token_interactive(
+        authority,
+        client_id,
+        account,
+        scopes,
+        prompt=None,  # TODO: Perhaps WAM would not accept this?
+        login_hint=None,  # type: Optional[str]
+        domain_hint=None,  # TODO: Perhaps WAM would not accept this?
+        claims_challenge=None,
+        timeout=None,  # TODO
+        extra_scopes_to_consent=None,  # TODO: Perhaps WAM would not accept this?
+        max_age=None,  # TODO: Perhaps WAM would not accept this?
+        **kwargs):
+    params = pymsalruntime.MSALRuntimeAuthParameters(client_id, authority)
+    params.set_requested_scopes(" ".join(scopes))
+    if login_hint:
+        params.set_login_hint(login_hint)
+    if claims_challenge:
+        params.set_claims(claims_challenge
+    # TODO: Wire up other input parameters too
+    callback_data = _CallbackData()
+    pymsalruntime.signin_interactive(
+        params,
+        "correlation", # TODO
+        account,
+        lambda result, callback_data=callback_data: callback_data.complete(result))
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+
+def acquire_token_interactive(
+        authority,  # type: str
+        client_id,  # type: str
+        scopes,  # type: list[str]
+        **kwargs):
+    """MSAL Python's acquire_token_interactive() will call this"""
+    result = _signin_silent(authority, client_id)
+    if not result.get_account():
+        result = _signin_interactive(authority, client_id)
+    if not result.get_account():
+        return {"error": result.get_error()}  # TODO
```

Comment:
nit: I am working on a sign_in API which you will be able to use that instead of silently followed by interactive. Thanks

Reply:
Good to know.

As you can see here, the boilerplate code here does not seem too overwhelming. But the bigger topic is that the draft here does not indicate a need to have 2 underlying functions: signin_silently() and signin_interactively(). That seems to be a mismatch of the different programming model. If our further experiments could indeed succeed with only one SignIn function, that would be very helpful in terms of simplifying the programming model.

Look forward to your new signin API.

Reply:
In my latest prototyping, I ended up attempting `signin_interactively()` *only* here. So, @MSamWils you might not need to work out a combined signin API at all. I'll let you know if our future iteration changes course.

Reply:
FYI, the latest pymsalruntime should have the signin API for you to integrate.